Add reinforce algorithm

SchwammDev opened the pull request:
Reinforce is one of the basic policy gradient algorithm. I've added it
to learn the basics of RL and policy gradient. So far it is just a first
draft and still missing crucial bits like policy approximations and
baseline estimators.

SchwammDev committed:
Add reinforce algorithm

Reinforce is one of the basic policy gradient algorithm. I've added it
to learn the basics of RL and policy gradient. So far it is just a first
draft and still missing crucial bits like policy approximations and
baseline estimators.

SchwammDev commented on reinforcement/trajectories.py:
```diff
@@ -0,0 +1,41 @@
+
+class Trajectories:
+    _RETURNS = 2
+    _ACTIONS = 1
```
These constants are obsolete

SchwammDev commented on tests/test_reinforce_algorithm.py:
```diff
@@ -0,0 +1,168 @@
+        CALL_ORDER.append(self.fit)
+        self.received_returns = list(batch.returns)
+        sim = OperationsSim()
+        losses = [self._calc(sim, a, r) for a, r in zip(batch.actions, batch.returns)]
```
Should pass in the whole batch (1/NM*Expectation)

Ermelyn replied:
Yes, thats right.

SchwammDev commented on reinforcement/algorithm/reinforce.py:
```diff
@@ -0,0 +1,36 @@
+
+
+def _calc_signal(ops, action_dist, returns):
+    return ops.reduce_mean(ops.log(action_dist) * returns)
```
This is probably wrong, the action probs need to be combined with the actual selected action in some way

SchwammDev replied:
This will be tackled once I've implemented general reinforcement learning tests. These should detect this issue reliably.

SchwammDev commented on tests/test_reinforce_algorithm.py:
```diff
@@ -0,0 +1,168 @@
+
+class BaselineStub:
+    def __init__(self, size):
+        self._estimates = np.random.uniform(size)
```
Size & random uniform not needed

SchwammDev replied:
Still necessary, but clarified naming

Ermelyn reviewed (changes requested):
Can't say more than you already mentioned.

SchwammDev committed:
Implement minor tweaks from review request

SchwammDev commented on tests/test_reinforce_algorithm.py:
```diff
@@ -0,0 +1,168 @@
+        CALL_ORDER.append(self.fit)
+        self.received_returns = list(batch.returns)
+        sim = OperationsSim()
+        losses = [self._calc(sim, a, r) for a, r in zip(batch.actions, batch.returns)]
```
Same as for the reward signal. The next merge should contain tests that hopefully can detect this issue reliably.

SchwammDev reviewed (commented):
Awaiting general reinforcement learning tests.

SchwammDev committed:
WIP: Add first learning tests

This is the start of implementing learning tests which should check on
trivial environments if an RL algorithm is learning something at all.
Additionally it forces me to create interfaces and modules for the other
components necessary for a rl algorithm to work.

Currently in the process of implementing it, but it is uploaded so I've
access to it on all my devices.

SchwammDev committed:
Add basic agent implementation & interface

To keep make it easier keeping different agents consistent an Agent
interface has been introduced. This interface acts as a contract that
each agent must fulfill to be used with the algorithms of this library.
Additionally a first batch agents has been implemented, which will be
used by the reinforce algorithm and the first tests.

SchwammDev committed:
Refactor trajectory recording

Cleaned up the piecewise recording of trajectories so it clearer how the
interface is supposed to be used. Also moved it into its own class so
there is a clear distinction between the different methods of creating
trajectories. Additionally, the TrajectoryBuilder has been moved to
the testing module as it is solely used by the automated tests.

SchwammDev committed:
Remove superfluous agents package folder

SchwammDev committed:
Implement basic learning test

Add a test that checks if an RL algorithm actually learns something on a
fairly simple environment. However, at the moment it is still WIP,
because the agent doesn't seem to be able to consistently beat the
random agent. There might still be a bug in the algorithm or test and
further investigation is required.

SchwammDev committed:
Implement first RL learning test

These automated tests using a most simple environment make sure that the
REINFORCE algorithm is actually learning something and converging
properly. They can also be used for fast diagnosis and understanding of
how the mechanics of RL work. Because the environment is so simple, a
fast iteration cycle and step by step debugging is possible. Still it
can shed light on many issues typically encountered when designing RL
algorithm (i.e. normalization errors). It also gives a feel on what
makes an algorithm stable and what not.

SchwammDev committed:
Update .gitignore to catch tensorboard logs

We don't want to commit tensorboard logs and the current system uses the
test name to distinguish between different tensorboard log runs.

SchwammDev committed:
Merge branch 'reinforce' into learning-tests

Conflicts:
	reinforcement/trajectories.py
	tests/test_reinforce_algorithm.py

SchwammDev committed:
Merge integration tests into tests folder

There was only one tests left in integration and it didn't even take
that long. We might however want to give the tests folder more structure
in the future. Also added a `__init__.py` to the tests folder to fix
missing coverage data. Also fixed some tensorflow v1 compability
warnings.

SchwammDev committed:
Use reruns for learning tests

It is in the nature of RL algorithms that depending on the
initialisation they might not learn an optimal solution. To account for
that we allow certain RL tests to be rerun again. This should make our
test suite more robust.

SchwammDev committed:
Implement review requests to improve code quality

Certain names are more clear now, and superfluous code has been deleted.
Additionally, the two main agent learning tests have been collapsed into
one, so it is easier to maintain in the future.

SchwammDev committed:
Ignore some lines for test coverage

By ignoring common lines such as NotImplementedErrors, ImportErrors,
main function checks etc. the test coverage measure becomes more
expressive. Additionally some missing tests have been added and some
superfluous code was detected.

SchwammDev committed:
Fix review requests

This makes the code clearer and more consistent.

SchwammDev committed:
Rename avg_total_return to avg_max_reward

This name better reflects, that it is the maximum achievable reward in
this environment, and is not confused by the average total return that
a random agent would achieve - which would be zero.

Commit:
Merge pull request #2 from SwamyDev/learning-tests

Implement first RL learning test

SchwammDev commented on tests/test_algorithm_learning.py:
```diff
@@ -0,0 +1,355 @@
+class LinearPolicy:
+    def __init__(self, plotter, lr=1):
+        self._lr = lr
+        self._m = np.random.normal(1, 0.001)  # high m avoids local maxima - use negative m for demonstration
```
```suggestion
        self._m = np.random.normal(1, 0.001)  # high slope avoids local maxima - use negative m for demonstration
```
maybe refer to m as slope in the comment to make it more clear

SchwammDev commented on tests/test_algorithm_learning.py:
```diff
@@ -0,0 +1,355 @@
+            fig, (mp, lp, rp) = plt.subplots(3, 1)
+            fig.suptitle("Numpy Agent")
+            mp.plot(self._m_history)
+            mp.set_title("m")
```
```suggestion
            mp.set_title("slope")
```
same here for the string

SchwammDev commented on tests/test_reinforce_algorithm.py:
```diff
@@ -0,0 +1,132 @@
+    assert CALL_ORDER == [baseline.estimate, policy.fit]
+
+
+def test_fitting_the_approximation_uses_the_correct_loss(policy, baseline, make_trajectory):
```
rename to reflect that it isn't loss signal anymore

SchwammDev reviewed (commented):
Fix changes

Commit:
Update comment to use slope instead of m

Clarifies what 'm' means

Commit:
Rename plot title from m to slope

Clarifies what 'm' means

SchwammDev committed:
Rename received_loss_signal to received_signal

Clarifies that this is not a loss yet, and that the sign has to be
determined by the policy approximation.

SchwammDev reviewed (commented):
Done

SchwammDev committed:
Complete renaming and avoid crash
